Performance improvements for dasherize.

dillonwelch opened the pull request:
Continuation of https://github.com/rails/rails/pull/30948

Add a cache to save the dasherized word so that we only generate the
string once.

This change prevents a string allocation for each time the method is
called, with a speed up of ~1.3x for small strings (5-10 chars) and
~2.4x for large strings (~50 chars)

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "benchmark-ips"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

module ActiveSupport
  module Inflector
    extend self
    def fast_dasherize(underscored_word)
      @_dasherize_cache ||= {}
      @_dasherize_cache[underscored_word] ||=
        if underscored_word.include?('_')
          underscored_word.tr('_', '-')
        else
          underscored_word
        end
    end
  end
end

%w(
  short_string
  a_medium_underscore_string
  really_long_string_with_many_underscores_omg_underscores_
  rails
).each do |value|
  puts " #{value} ".center(60, '=')

  puts " Memory Usage ".center(60, "=")
  puts

  puts "value.dasherize"
  puts allocate_count {
    1000.times { ActiveSupport::Inflector.dasherize(value) }
  }

  puts "value.fast_dasherize"
  puts allocate_count {
    1000.times { ActiveSupport::Inflector.fast_dasherize(value) }
  }

  puts
  puts " Benchmark.ips ".center(60, "=")
  puts

  Benchmark.ips do |x|
    x.report("dasherize") do
      ActiveSupport::Inflector.dasherize(value)
    end
    x.report("fast_dasherize") do
      ActiveSupport::Inflector.fast_dasherize(value)
    end
    x.compare!
  end
end
```

```
======================= short_string =======================
======================= Memory Usage =======================

value.dasherize
{:FREE=>-988, :T_STRING=>1052, :T_IMEMO=>1}
value.fast_dasherize
{:FREE=>-8, :T_STRING=>5, :T_HASH=>1, :T_IMEMO=>1}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize   148.831k i/100ms
      fast_dasherize   127.826k i/100ms
Calculating -------------------------------------
           dasherize      2.949M (±15.7%) i/s -     14.437M in   5.042609s
      fast_dasherize      3.672M (±12.6%) i/s -     18.023M in   5.005492s

Comparison:
      fast_dasherize:  3672161.3 i/s
           dasherize:  2948709.3 i/s - same-ish: difference falls within error

================ a_medium_underscore_string ================
======================= Memory Usage =======================

value.dasherize
{:FREE=>-1001, :T_STRING=>1000}
value.fast_dasherize
{:FREE=>-5, :T_STRING=>4}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize   107.950k i/100ms
      fast_dasherize   159.392k i/100ms
Calculating -------------------------------------
           dasherize      1.676M (± 8.5%) i/s -      8.420M in   5.058826s
      fast_dasherize      3.539M (± 8.9%) i/s -     17.533M in   5.000014s

Comparison:
      fast_dasherize:  3538794.2 i/s
           dasherize:  1676062.3 i/s - 2.11x  slower

 really_long_string_with_many_underscores_omg_underscores_ =
======================= Memory Usage =======================

value.dasherize
{:FREE=>-1001, :T_STRING=>1000}
value.fast_dasherize
{:FREE=>-5, :T_STRING=>4}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize    96.655k i/100ms
      fast_dasherize   148.701k i/100ms
Calculating -------------------------------------
           dasherize      1.416M (±18.7%) i/s -      6.863M in   5.048829s
      fast_dasherize      3.464M (±10.0%) i/s -     17.249M in   5.037720s

Comparison:
      fast_dasherize:  3463915.3 i/s
           dasherize:  1416219.4 i/s - 2.45x  slower

========================== rails ===========================
======================= Memory Usage =======================

value.dasherize
{:FREE=>-1001, :T_STRING=>1000}
value.fast_dasherize
{:FREE=>-3, :T_STRING=>2}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize   157.250k i/100ms
      fast_dasherize   158.780k i/100ms
Calculating -------------------------------------
           dasherize      3.348M (± 7.8%) i/s -     16.669M in   5.013025s
      fast_dasherize      3.855M (±10.3%) i/s -     19.054M in   5.005788s

Comparison:
      fast_dasherize:  3855075.6 i/s
           dasherize:  3347527.7 i/s - same-ish: difference falls within error
```

dillonwelch committed:
Performance improvements for dasherize.

Add a cache to save the dasherized word so that we only generate the
string once.

This change prevents a string allocation for each time the method is
called, with a speed up of ~1.3x for small strings (5-10 chars) and
~2.4x for large strings (~50 chars)

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "benchmark-ips"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

module ActiveSupport
  module Inflector
    extend self
    def fast_dasherize(underscored_word)
      @_dasherize_cache ||= {}
      @_dasherize_cache[underscored_word] ||=
        if underscored_word.include?('_')
          underscored_word.tr('_', '-')
        else
          underscored_word
        end
    end
  end
end

puts
puts " dasherize ".center(60, "=")
puts

%w(
  short_string
  a_medium_underscore_string
  really_long_string_with_many_underscores_omg_underscores_
  rails
).each do |value|
  puts " #{value} ".center(60, '=')

  puts " Memory Usage ".center(60, "=")
  puts

  puts "value.dasherize"
  puts allocate_count {
    1000.times { ActiveSupport::Inflector.dasherize(value) }
  }

  puts "value.fast_dasherize"
  puts allocate_count {
    1000.times { ActiveSupport::Inflector.fast_dasherize(value) }
  }

  puts
  puts " Benchmark.ips ".center(60, "=")
  puts

  Benchmark.ips do |x|
    x.report("dasherize") do
      ActiveSupport::Inflector.dasherize(value)
    end
    x.report("fast_dasherize") do
      ActiveSupport::Inflector.fast_dasherize(value)
    end
    x.compare!
  end
end
======================= short_string =======================
======================= Memory Usage =======================

value.dasherize
{:FREE=>-988, :T_STRING=>1052, :T_IMEMO=>1}
value.fast_dasherize
{:FREE=>-8, :T_STRING=>5, :T_HASH=>1, :T_IMEMO=>1}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize   148.831k i/100ms
      fast_dasherize   127.826k i/100ms
Calculating -------------------------------------
           dasherize      2.949M (±15.7%) i/s -     14.437M in   5.042609s
      fast_dasherize      3.672M (±12.6%) i/s -     18.023M in   5.005492s

Comparison:
      fast_dasherize:  3672161.3 i/s
           dasherize:  2948709.3 i/s - same-ish: difference falls within error

================ a_medium_underscore_string ================
======================= Memory Usage =======================

value.dasherize
{:FREE=>-1001, :T_STRING=>1000}
value.fast_dasherize
{:FREE=>-5, :T_STRING=>4}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize   107.950k i/100ms
      fast_dasherize   159.392k i/100ms
Calculating -------------------------------------
           dasherize      1.676M (± 8.5%) i/s -      8.420M in   5.058826s
      fast_dasherize      3.539M (± 8.9%) i/s -     17.533M in   5.000014s

Comparison:
      fast_dasherize:  3538794.2 i/s
           dasherize:  1676062.3 i/s - 2.11x  slower

 really_long_string_with_many_underscores_omg_underscores_ =
======================= Memory Usage =======================

value.dasherize
{:FREE=>-1001, :T_STRING=>1000}
value.fast_dasherize
{:FREE=>-5, :T_STRING=>4}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize    96.655k i/100ms
      fast_dasherize   148.701k i/100ms
Calculating -------------------------------------
           dasherize      1.416M (±18.7%) i/s -      6.863M in   5.048829s
      fast_dasherize      3.464M (±10.0%) i/s -     17.249M in   5.037720s

Comparison:
      fast_dasherize:  3463915.3 i/s
           dasherize:  1416219.4 i/s - 2.45x  slower

========================== rails ===========================
======================= Memory Usage =======================

value.dasherize
{:FREE=>-1001, :T_STRING=>1000}
value.fast_dasherize
{:FREE=>-3, :T_STRING=>2}

====================== Benchmark.ips =======================

Warming up --------------------------------------
           dasherize   157.250k i/100ms
      fast_dasherize   158.780k i/100ms
Calculating -------------------------------------
           dasherize      3.348M (± 7.8%) i/s -     16.669M in   5.013025s
      fast_dasherize      3.855M (±10.3%) i/s -     19.054M in   5.005788s

Comparison:
      fast_dasherize:  3855075.6 i/s
           dasherize:  3347527.7 i/s - same-ish: difference falls within error

jeremy reviewed (commented):
`params[:foo].dasherize` → user input leads to growing cache size until it exhausts all memory, causing denial of service?

rafaelfranca commented:
> params[:foo].dasherize → user input leads to growing cache size until it exhausts all memory, causing denial of service?

Yes.

oniofchaos commented on activesupport/lib/active_support/inflector/methods.rb:
```diff
@@ -207,7 +207,13 @@ def classify(table_name)
-      underscored_word.tr("_".freeze, "-".freeze)
+      @_dasherize_cache ||= {}
+      @_dasherize_cache[underscored_word] ||=
+        if underscored_word.include?("_")
```
@rafaelfranca @jeremy thoughts on this change separate from the caching?

rafaelfranca replied:
Looks fine to me.

oniofchaos replied:
Did some profiling on this and in the scenario with no underscore it provides a speedup but in the scenario with an underscore it was equivalently slower - which doesn't seem worth it to me.

oniofchaos commented:
The use case where I discovered this was when rendering a link with a `disable_with` argument for each model on an index page. Given the security issue that an unbounded cache that can store user input presents, do you have suggestions for another bit of code to change that would lead to similar results (not dasherizing the same string over and over). 

Also, I'm testing this on Rails 4.2.10 (version my app is on), perhaps this is fixed/alleviated in master?

```haml
- @results.each do |model|
  %td= model.name
  %td= link_to(t('.delete'), model_path(model), { data: { disable_with: 'Please wait...' }, method: :delete }
```

Caller trace:
```ruby
[193, 202] in /Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/inflector/methods.rb
   193:       #                                         end
   194: 
   195:       # byebug unless underscored_word.include?('_'.freeze)
   196:       if underscored_word.include?('_'.freeze)
   197:         byebug
=> 198:         underscored_word.tr('_'.freeze, '-'.freeze)
   199:       else
   200:         # @_count ||= 0
   201:         # @_count += 1
   202:         # puts "count is now #{@_count} because of  #{underscored_word}"
(byebug) underscored_word
"disable_with"
(byebug) puts caller
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/helpers/eval.rb:59:in `eval'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/helpers/eval.rb:59:in `safe_eval'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/helpers/eval.rb:53:in `warning_eval'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/helpers/eval.rb:29:in `block in multiple_thread_eval'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/helpers/eval.rb:92:in `allowing_other_threads'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/helpers/eval.rb:29:in `multiple_thread_eval'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/processors/command_processor.rb:164:in `block in run_cmd'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/processors/command_processor.rb:169:in `safely'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/processors/command_processor.rb:160:in `run_cmd'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/processors/command_processor.rb:134:in `repl'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/processors/command_processor.rb:98:in `process_commands'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/processors/command_processor.rb:56:in `at_line'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/byebug-9.1.0/lib/byebug/context.rb:96:in `at_line'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/inflector/methods.rb:198:in `dasherize'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/core_ext/string/inflections.rb:126:in `dasherize'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:167:in `prefix_tag_option'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:155:in `block (2 levels) in tag_options'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:154:in `each_pair'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:154:in `block in tag_options'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:152:in `each_pair'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:152:in `tag_options'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:144:in `content_tag_string'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/tag_helper.rb:108:in `content_tag'
/Users/dillonwelch/Code/q-centrix/haml/lib/haml/helpers/action_view_mods.rb:66:in `content_tag_with_haml'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/url_helper.rb:184:in `link_to'
/Users/dillonwelch/Code/q-centrix/web/app/views/admin/patients/index.html.haml:4:in `block in _app_views_admin_patients_index_html_haml__4483526660187390292_70093878372920'
/Users/dillonwelch/Code/q-centrix/haml/lib/haml/helpers.rb:382:in `block in capture_haml'
/Users/dillonwelch/Code/q-centrix/haml/lib/haml/helpers.rb:667:in `with_haml_buffer'
/Users/dillonwelch/Code/q-centrix/haml/lib/haml/helpers.rb:378:in `capture_haml'
/Users/dillonwelch/Code/q-centrix/haml/lib/haml/helpers/xss_mods.rb:62:in `capture_haml_with_haml_xss'
/Users/dillonwelch/Code/q-centrix/haml/lib/haml/helpers/action_view_mods.rb:45:in `capture_with_haml'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/helpers/capture_helper.rb:152:in `content_for'
/Users/dillonwelch/Code/q-centrix/web/app/views/admin/patients/index.html.haml:1:in `_app_views_admin_patients_index_html_haml__4483526660187390292_70093878372920'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/template.rb:145:in `block in render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications.rb:166:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/template.rb:333:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/template.rb:143:in `render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-mini-profiler-0.10.5/lib/mini_profiler/profiling_methods.rb:76:in `block in profile_method'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/template_renderer.rb:54:in `block (2 levels) in render_template'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/abstract_renderer.rb:39:in `block in instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications.rb:164:in `block in instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications.rb:164:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/abstract_renderer.rb:39:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/template_renderer.rb:53:in `block in render_template'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/template_renderer.rb:61:in `render_with_layout'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/template_renderer.rb:52:in `render_template'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/template_renderer.rb:14:in `render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/renderer.rb:46:in `render_template'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/renderer/renderer.rb:27:in `render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/rendering.rb:100:in `_render_template'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/streaming.rb:217:in `_render_template'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/rendering.rb:83:in `render_to_body'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/rendering.rb:32:in `render_to_body'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/renderers.rb:37:in `render_to_body'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/abstract_controller/rendering.rb:25:in `render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/rendering.rb:16:in `render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/instrumentation.rb:44:in `block (2 levels) in render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/core_ext/benchmark.rb:12:in `block in ms'
/Users/dillonwelch/.rvm/rubies/ruby-2.4.2/lib/ruby/2.4.0/benchmark.rb:308:in `realtime'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/core_ext/benchmark.rb:12:in `ms'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/instrumentation.rb:44:in `block in render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/instrumentation.rb:87:in `cleanup_view_runtime'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activerecord-4.2.10/lib/active_record/railties/controller_runtime.rb:25:in `cleanup_view_runtime'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/instrumentation.rb:43:in `render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/wicked_pdf-1.1.0/lib/wicked_pdf/pdf_helper.rb:42:in `render_with_wicked_pdf'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/implicit_render.rb:10:in `default_render'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/implicit_render.rb:5:in `send_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/abstract_controller/base.rb:198:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/rendering.rb:10:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:117:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:505:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:498:in `block (2 levels) in around'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:292:in `block (2 levels) in halting_and_conditional'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/core_ext/time/zones.rb:48:in `use_zone'
/Users/dillonwelch/Code/q-centrix/web/app/controllers/application_controller.rb:79:in `set_time_zone'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:432:in `block in make_lambda'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:291:in `block in halting_and_conditional'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:497:in `block in around'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:505:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:498:in `block (2 levels) in around'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:292:in `block (2 levels) in halting_and_conditional'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:432:in `block in make_lambda'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:291:in `block in halting_and_conditional'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:497:in `block in around'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:505:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:92:in `__run_callbacks__'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:778:in `_run_process_action_callbacks'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:81:in `run_callbacks'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/abstract_controller/callbacks.rb:19:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/rescue.rb:29:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/instrumentation.rb:32:in `block in process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications.rb:164:in `block in instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/notifications.rb:164:in `instrument'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/instrumentation.rb:30:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activerecord-4.2.10/lib/active_record/railties/controller_runtime.rb:18:in `process_action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/abstract_controller/base.rb:137:in `process'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionview-4.2.10/lib/action_view/rendering.rb:30:in `process'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-mini-profiler-0.10.5/lib/mini_profiler/profiling_methods.rb:76:in `block in profile_method'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal.rb:196:in `dispatch'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_controller/metal.rb:237:in `block in action'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/routing/route_set.rb:74:in `dispatch'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/routing/route_set.rb:43:in `serve'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/journey/router.rb:43:in `block in serve'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/journey/router.rb:30:in `each'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/journey/router.rb:30:in `serve'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/routing/route_set.rb:817:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/warden-1.2.7/lib/warden/manager.rb:36:in `block in call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/warden-1.2.7/lib/warden/manager.rb:35:in `catch'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/warden-1.2.7/lib/warden/manager.rb:35:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/etag.rb:24:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/conditionalget.rb:25:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/head.rb:13:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/params_parser.rb:27:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/flash.rb:260:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/session/abstract/id.rb:225:in `context'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/session/abstract/id.rb:220:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/cookies.rb:560:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activerecord-4.2.10/lib/active_record/query_cache.rb:36:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/connection_pool.rb:653:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:88:in `__run_callbacks__'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:778:in `_run_call_callbacks'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/callbacks.rb:81:in `run_callbacks'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/railties-4.2.10/lib/rails/rack/logger.rb:38:in `call_app'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/railties-4.2.10/lib/rails/rack/logger.rb:20:in `block in call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/tagged_logging.rb:68:in `block in tagged'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/tagged_logging.rb:26:in `tagged'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/activesupport-4.2.10/lib/active_support/tagged_logging.rb:68:in `tagged'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/railties-4.2.10/lib/rails/rack/logger.rb:20:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/request_store-1.3.2/lib/request_store/middleware.rb:9:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/request_id.rb:21:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/methodoverride.rb:22:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/runtime.rb:18:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/actionpack-4.2.10/lib/action_dispatch/middleware/static.rb:120:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/sendfile.rb:113:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-mini-profiler-0.10.5/lib/mini_profiler/profiler.rb:212:in `block in call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/memory_profiler-0.9.8/lib/memory_profiler/reporter.rb:67:in `run'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/memory_profiler-0.9.8/lib/memory_profiler/reporter.rb:31:in `report'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/memory_profiler-0.9.8/lib/memory_profiler.rb:13:in `report'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-mini-profiler-0.10.5/lib/mini_profiler/profiler.rb:211:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/railties-4.2.10/lib/rails/engine.rb:518:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/railties-4.2.10/lib/rails/application.rb:165:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/lock.rb:17:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/content_length.rb:15:in `call'
/Users/dillonwelch/.rvm/gems/ruby-2.4.2/gems/rack-1.6.8/lib/rack/handler/webrick.rb:88:in `service'
/Users/dillonwelch/.rvm/rubies/ruby-2.4.2/lib/ruby/2.4.0/webrick/httpserver.rb:140:in `service'
/Users/dillonwelch/.rvm/rubies/ruby-2.4.2/lib/ruby/2.4.0/webrick/httpserver.rb:96:in `run'
/Users/dillonwelch/.rvm/rubies/ruby-2.4.2/lib/ruby/2.4.0/webrick/server.rb:290:in `block in start_thread'
```

sobrinho commented:
I'm not sure if it worth but you may try to use a private dict of common attrs like this:

```ruby
CACHE = {
  'disable_with' => 'disable-with'
}.freeze
```

In that way, we could improve the most common dasherization without an unbound cache.

In the other hand, we could use a bounded MRU cache or something like that to keep it generic to cover the entire application or even maybe a combination of both.

dillonwelch committed:
Move performance change to prefix_tag_option.

Add logic to check if the key/value pair is a common option, such as
data and disable with. For these common options, we can skip the
dasherize and stringify calls directly and have the finalized string
prepared.

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "benchmark-ips"
  gem "rails"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

def master_version
  key = "#{@prefix}-#{@key.to_s.dasherize}"
end

def fast_version
  if @prefix == 'data' and @key == :disable_with
    key = "data-disable-with"
  else
    key = "#{@prefix}-#{@key.to_s.dasherize}"
  end
end

[['data', :disable_with], ['data', :some_other_key]].each do |opt|
  @prefix = opt[0]
  @key = opt[1]

  puts "#{@prefix} #{@key}"
  puts "master_version"
  puts allocate_count { 1000.times { master_version } }
  puts "fast_version"
  puts allocate_count { 1000.times { fast_version } }

  Benchmark.ips do |x|
    x.report("master_version") { master_version }
    x.report("fast_version")     { fast_version }
    x.compare!
  end
end
```

```ruby
data disable_with
master_version
{:FREE=>-2826, :T_STRING=>3052, :T_IMEMO=>1}
fast_version
{:FREE=>-1}
Warming up --------------------------------------
      master_version    64.007k i/100ms
        fast_version   104.339k i/100ms
Calculating -------------------------------------
      master_version      1.404M (± 9.0%) i/s -      6.977M in   5.014783s
        fast_version      6.188M (±15.0%) i/s -     30.050M in   5.009728s

Comparison:
        fast_version:  6188197.7 i/s
      master_version:  1403550.9 i/s - 4.41x  slower

data some_other_key
master_version
{:FREE=>-3001, :T_STRING=>3000}
fast_version
{:FREE=>-3001, :T_STRING=>3000}
Warming up --------------------------------------
      master_version    67.354k i/100ms
        fast_version    78.323k i/100ms
Calculating -------------------------------------
      master_version      1.373M (± 8.0%) i/s -      6.870M in   5.036435s
        fast_version      1.190M (±11.4%) i/s -      5.953M in   5.076867s

Comparison:
      master_version:  1372924.9 i/s
        fast_version:  1190217.0 i/s - same-ish: difference falls within error
```

dillonwelch committed:
Revert change to dasherize

oniofchaos commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -95,7 +95,11 @@ def tag_option(key, value, escape)
           def prefix_tag_option(prefix, key, value, escape)
-            key = "#{prefix}-#{key.to_s.dasherize}"
+            if prefix == 'data' and key == :disable_with
+              key = "data-disable-with"
```
See the full details in https://github.com/rails/rails/pull/30962/commits/a3f4e8a607bdef8870ed043f1981e03f957ae354 but essentially this provides a 4.41x speedup for strings that quality.

oniofchaos commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -95,7 +95,11 @@ def tag_option(key, value, escape)
         private
           def prefix_tag_option(prefix, key, value, escape)
-            key = "#{prefix}-#{key.to_s.dasherize}"
+            if prefix == 'data' and key == :disable_with
```
The page I've been profiling on only has this option. @rafaelfranca do you know how to find other common options to add to this?

dillonwelch committed:
Fix rubocop style issues

rafaelfranca commented:
Is the difference in a real scenario, not in a micro benchmark that much to justify increasing the complexity of a method?

oniofchaos commented:
It impacts every call to `link_to` that has a `{data { disable_with: 'string' } }` option passed to it. I have a page in my app that can return hundreds/thousands of results (probably need pagination!) and each result has three associated links that each have the disable option.  Likely minor in the grand scheme of performance (though a strong 4.4x improvement at the micro level) but it does save potentially MBs/hundreds of KBs for pages with many links.

rafaelfranca commented:
We don't drive performance improvements in the framework using micro benchmarks. If it is not noticeable in the grand schema of performance it is not worth the extra complexity we are adding to the codebase.

It looks simple but now we have a conditional to a specific argument of that method. And why only that possible value and not all the other possible values.

That being said, even that it may save some MBs for some pages, I don't think it is worth to add all this complexity in the codebase.

If the performance of this method is sensitive to you, you could avoid it all together doing: `"data-disable-with" => "string"` in your application.

oniofchaos commented:
> And why only that possible value and not all the other possible values.

I'd love to add other arguments but I wasn't aware of others. If you know of others I can add them and then move this into a constant instead. Something like
```ruby
COMMON_TAG_OPTIONS = {
  disable_with: 'data-disable-with',
  other_option: 'data-other-option',
  ...
}.freeze

if prefix == "data" and COMMON_TAG_OPTIONS.key?(key)
  key = COMMON_TAG_OPTIONS[key]
else
  key = "#{prefix}-#{key.to_s.dasherize}"
end
```
or if you want to hide it in a method
```ruby
key = key_for_common_tags(prefix, key) || "#{prefix}-#{key.to_s.dasherize}"
```

> If the performance of this method is sensitive to you, you could avoid it all together doing: "data-disable-with" => "string" in your application.

Woah, I did not know about this option and after testing it does appear to help. Is there a reason it's not listed as the recommended/alternative option in docs? http://api.rubyonrails.org/classes/ActionView/Helpers/UrlHelper.html#method-i-link_to
